### ISSUE TITLE
Don't add parameters after functions 

### ISSUE THREAD

**Issue by fregante** [sindresorhus/memoize#97]
While this reads fine:

```js
const addOneListener = memoize(addListener, {
	cacheKey: arguments_ => arguments_,
	cache: new ManyKeysMap()
});
```

This is bad practice:

```js
const addOneListener = memoize(() => {
	// some


	// inline


	// long


	// function


	// which


	// is not


	// uncommon
	return stuff
}), {
	cacheKey: arguments_ => arguments_,
	cache: new ManyKeysMap()
});
```

This could be enforced by a linter, since APIs like `addEventListener(string, cb, options)` exist natively, but ideally `memoize` and `p-memoize` should accept a single parameter:

```js
function memoize(details: Callback | Options) {
	const {callback, ...options} = typeof details === 'function' ? {callback: details} : details;
}
```

It complicates types a bit but it requires the much cleaner:

```js
const addOneListener = memoize({
	cacheKey: arguments_ => arguments_,
	cache: new ManyKeysMap(),
	callback: () => {
		// some
	
	
		// inline
	
	
		// long
	
	
		// function
	
	
		// which
	
	
		// is not
	
	
		// uncommon
		return stuff
	}),
});
```

--- 
_Personal note:_ I went through this [recently in webext-storage-cache](https://github.com/fregante/webext-storage-cache/pull/41), changing `(fn, opts) => {}` to `(fnOrOpts) => {}`

**Comment by sindresorhus** [comment]
I don't see why the API has to change. Just not put the function inline.

**Comment by sindresorhus** [comment]
> but ideally memoize and p-memoize should accept a single parameter:

I disagree about that being ideal. I find that API quite weird. There are now two ways to do the same thing.

**Comment by sindresorhus** [comment]
> Should I open an issue on unicorn?

👍

**Comment by fregante** [comment]
> I find that API quite weird. There are now two ways to do the same thing.

have you seen fetch 🙈 

```js
fetch('flowers.jpg', {method: 'POST'})
```
```js
const myRequest = new Request("flowers.jpg", {method: 'POST'});
fetch(myRequest)
```


> two ways to do the same thing.

I mean, that's pretty common, but I don’t think anyone would prefer `mem({cb: () => {}})` over `mem(() => {})` unless the API requires it (for good reasons, as shown here)

**Comment by fisker** [comment]
Maybe we can change signature to

`memoize(options?, () => {})` 

?

options can't be a function anyway.

**Comment by fregante** [comment]
That sounds almost good but I think `(options, callback)` is not a common pattern

**Comment by fisker** [comment]
`(Callback | Options)` is not diff friendly

```diff
memoize({
-   cb() {
-     // .... indention level 2
-   },
-   cache: new ManyKeysMap,
})
```

```diff
memoize(() => {
+   // .... indention level 1
})
```

**Comment by meduzen** [comment]
> Inline functions are nice but they grow and people don't move them out.

They do when you tell them. :p

**Comment by fregante** [comment]
> > Inline functions are nice but they grow and people don't move them out.
> 
> They do when you tell them. :p

Sounds like a linter job, but that rule is complex and I think it's not in the recommended config, so it's easier to just avoid bad patterns _at the root_

**Comment by fregante** [comment]
> `(Callback | Options)` is not diff friendly

Lots of things are not diff friendly, like Prettier, and yet… they're worth it.

**Comment by sindresorhus** [comment]
I'm going to pass on this. I don't want two ways to do something and not worth a breaking change. But more importantly, I think the current API is fine.
